pkg/phlaredb/bucketindex/index.go

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/pkg/storage/tsdb/bucketindex/index.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Cortex Authors.
+
+package bucketindex
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/oklog/ulid"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/pyroscope/pkg/phlaredb/block"
+	"github.com/grafana/pyroscope/pkg/phlaredb/sharding"
+	"github.com/grafana/pyroscope/pkg/util"
+)
+
+const (
+	IndexFilename           = "bucket-index.json"
+	IndexCompressedFilename = IndexFilename + ".gz"
+	IndexVersion1           = 1
+	IndexVersion2           = 2 // Added CompactorShardID field.
+	SegmentsFormatUnknown   = ""
+
+	// SegmentsFormat1Based6Digits defined segments numbered with 6 digits numbers in a sequence starting from number 1
+	// eg. (000001, 000002, 000003).
+	SegmentsFormat1Based6Digits = "1b6d"
+)
+
+// Index contains all known blocks and markers of a tenant.
+type Index struct {
+	// Version of the index format.
+	Version int `json:"version"`
+
+	// List of complete blocks (partial blocks are excluded from the index).
+	Blocks Blocks `json:"blocks"`
+
+	// List of block deletion marks.
+	BlockDeletionMarks BlockDeletionMarks `json:"block_deletion_marks"`
+
+	// UpdatedAt is a unix timestamp (seconds precision) of when the index has been updated
+	// (written in the storage) the last time.
+	UpdatedAt int64 `json:"updated_at"`
+}
+
+func (idx *Index) GetUpdatedAt() time.Time {
+	return time.Unix(idx.UpdatedAt, 0)
+}
+
+// RemoveBlock removes block and its deletion mark (if any) from index.
+func (idx *Index) RemoveBlock(id ulid.ULID) {
+	for i := 0; i < len(idx.Blocks); i++ {
+		if idx.Blocks[i].ID == id {
+			idx.Blocks = append(idx.Blocks[:i], idx.Blocks[i+1:]...)
+			break
+		}
+	}
+
+	for i := 0; i < len(idx.BlockDeletionMarks); i++ {
+		if idx.BlockDeletionMarks[i].ID == id {
+			idx.BlockDeletionMarks = append(idx.BlockDeletionMarks[:i], idx.BlockDeletionMarks[i+1:]...)
+			break
+		}
+	}
+}
```

Comment:
A nit: an object is still retained when removed from `idx.Blocks` and `idx.BlockDeletionMarks`, we should nil it